arch/posix/core/posix_core.c

```diff
@@ -0,0 +1,556 @@
+ * independently as possible to the remainder of Zephyr to avoid name clashes
+ * as Zephyr does provide functions with the same names as the POSIX threads
+ * functions
+ */
+/**
+ * Principle of operation:
+ *
+ * The Zephyr OS and its app run as a set of native pthreads.
+ * The Zephyr OS only sees one of this thread executing at a time.
+ * Which is running is controlled using pc_{cond|mtx}_threads and
+ * pc_currently_allowed_thread.
+ *
+ * The main part of the execution of each thread will occur in a fully
+ * synchronous and deterministic manner, and only when commanded by the Zephyr
+ * kernel.
+ * But the creation of a thread will spawn a new pthread whose start
+ * is asynchronous to the rest, until synchronized in pc_wait_until_allowed()
+ * below.
+ * Similarly aborting and canceling threads execute a tail in a quite
+ * asynchronous manner.
+ *
+ * This implementation is meant to be portable in between POSIX systems.
+ * A table (pc_threads_table) is used to abstract the native pthreads.
+ * And index in this table is used to identify threads in the IF to the kernel.
+ *
+ */
+
+#define POSIX_ARCH_DEBUG_PRINTS 0
+
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "posix_core.h"
+#include "posix_soc_if.h"
+#include "nano_internal.h"
+#include "kernel_structs.h"
+#include "ksched.h"
+
+#define PREFIX     "POSIX arch core: "
+#define ERPREFIX   PREFIX"error on "
+#define NO_MEM_ERR PREFIX"Can't allocate memory\n"
+
+#define PC_ALLOC_CHUNK_SIZE 64
+#define PC_REUSE_ABORTED_ENTRIES 0
+/* tests/kernel/threads/scheduling/schedule_api fails when setting
+ * PC_REUSE_ABORTED_ENTRIES => don't set it by now
+ */
+
+static int pc_threads_table_size;
+typedef struct {
+	enum {NotUsed = 0, Used, Aborting, Aborted, Failed} state;
+	bool running;     /*is this the currently running thread*/
+	pthread_t thread; /*actual pthread_t as returned by native kernel*/
+	int thead_cnt; /*for debugging. Unique, consecutive, thread number*/
+} pc_threads_table_t;
+
+static pc_threads_table_t *pc_threads_table;
+
+static int thread_create_count; /*For debugging. Thread creation counter*/
+
+/*
+ * Conditional variable to block/awake all threads during swaps()
+ * (we only need 1 mutex and 1 cond variable for all threads)
+ */
+static pthread_cond_t pc_cond_threads = PTHREAD_COND_INITIALIZER;
+/* Mutex for the conditional variable posix_core_cond_threads */
+static pthread_mutex_t pc_mtx_threads = PTHREAD_MUTEX_INITIALIZER;
+/* Token which tells which process is allowed to run now */
+static int pc_currently_allowed_thread;
+
+static bool pc_terminate; /*Are we terminating the program == cleaning up*/
+
+static void pc_wait_until_allowed(int this_th_nbr);
+static void *pc_thread_starter(void *arg);
+static void pc_preexit_cleanup(void);
+
+/**
+ * Helper function, run by a thread is being aborted
+ */
+static void pc_abort_tail(int this_th_nbr)
+{
+	if (POSIX_ARCH_DEBUG_PRINTS) {
```

Comment:
Use a macro instead of if statements in all these cases, similar to: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/common/log.h#L40

Reply:
Makes sense. There is a lot of spurious code for this prints